File: service-beacons-go/services/version1/BeaconsRestServiceV1.go
package services1

import (
	"encoding/json"
	"io/ioutil"
	"net/http"
	"strings"

	"github.com/gorilla/mux"
	data1 "github.com/nov-pocs/samples/service-beacons-go/data/version1"
	logic "github.com/nov-pocs/samples/service-beacons-go/logic"
	cconv "github.com/pip-services3-go/pip-services3-commons-go/convert"
	cdata "github.com/pip-services3-go/pip-services3-commons-go/data"
	cerr "github.com/pip-services3-go/pip-services3-commons-go/errors"
	crefer "github.com/pip-services3-go/pip-services3-commons-go/refer"
	cvalid "github.com/pip-services3-go/pip-services3-commons-go/validate"
	cservices "github.com/pip-services3-go/pip-services3-rpc-go/services"
)

type BeaconsRestService struct {
	*cservices.RestService
	controller    logic.IBeaconsController
	numberOfCalls int
}

func NewBeaconsRestService() *BeaconsRestService {
	c := &BeaconsRestService{}
	c.RestService = cservices.NewRestService()
	c.RestService.IRegisterable = c
	c.numberOfCalls = 0
	c.BaseRoute = "v1/beacons"
	c.DependencyResolver.Put("controller", crefer.NewDescriptor("beacons", "controller", "default", "*", "*"))
	return c
}

func (c *BeaconsRestService) SetReferences(references crefer.IReferences) {
	c.RestService.SetReferences(references)
	ctrl, err := c.DependencyResolver.GetOneRequired("controller")
	if err == nil && ctrl != nil {
		c.controller = ctrl.(logic.IBeaconsController)
	}
}

func (c *BeaconsRestService) getPageByFilter(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	pagingParams := make(map[string]string, 0)

	pagingParams["skip"] = params.Get("skip")
	pagingParams["take"] = params.Get("take")
	pagingParams["total"] = params.Get("total")

	delete(params, "skip")
	delete(params, "take")
	delete(params, "total")

	result, err := c.controller.GetBeacons(
		params.Get("correlation_id"),
		cdata.NewFilterParamsFromValue(params), // W! need test
		cdata.NewPagingParamsFromValue(pagingParams),
	)
	c.SendResult(res, req, result, err)
}

func (c *BeaconsRestService) getOneById(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	vars := mux.Vars(req)

	beaconId := params.Get("beacon_id")
	if beaconId == "" {
		beaconId = vars["beacon_id"]
	}

	result, err := c.controller.GetBeaconById(
		params.Get("correlation_id"),
		beaconId)
	c.SendResult(res, req, result, err)
}

func (c *BeaconsRestService) getOneByUdi(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	vars := mux.Vars(req)

	beaconUdi := params.Get("udi")
	if beaconUdi == "" {
		beaconUdi = vars["udi"]
	}

	result, err := c.controller.GetBeaconByUdi(
		params.Get("correlation_id"),
		beaconUdi)
	c.SendResult(res, req, result, err)
}

func (c *BeaconsRestService) calculatePosition(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()

	udis := params.Get("udis")
	udiList := make([]string, 0, 0)
	udiList = strings.Split(udis, ",")

	siteId := params.Get("site_id")

	result, err := c.controller.CalculatePosition(
		params.Get("correlation_id"),
		siteId,
		udiList)
	c.SendResult(res, req, result, err)
}

func (c *BeaconsRestService) create(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	correlationId := params.Get("correlation_id")
	var beacon data1.BeaconV1

	body, bodyErr := ioutil.ReadAll(req.Body)
	if bodyErr != nil {
		err := cerr.NewInternalError(correlationId, "JSON_CNV_ERR", "Can't convert from JSON to Beacons").WithCause(bodyErr)
		c.SendError(res, req, err)
		return
	}
	defer req.Body.Close()
	jsonErr := json.Unmarshal(body, &beacon)

	if jsonErr != nil {
		err := cerr.NewInternalError(correlationId, "JSON_CNV_ERR", "Can't convert from JSON to Beacons").WithCause(jsonErr)
		c.SendError(res, req, err)
		return
	}

	result, err := c.controller.CreateBeacon(
		correlationId,
		&beacon,
	)
	c.SendCreatedResult(res, req, result, err)
}

func (c *BeaconsRestService) update(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	correlationId := params.Get("correlation_id")

	var beacon data1.BeaconV1

	body, bodyErr := ioutil.ReadAll(req.Body)
	if bodyErr != nil {
		err := cerr.NewInternalError(correlationId, "JSON_CNV_ERR", "Can't convert from JSON to Beacons").WithCause(bodyErr)
		c.SendError(res, req, err)
		return
	}
	defer req.Body.Close()
	jsonErr := json.Unmarshal(body, &beacon)

	if jsonErr != nil {
		err := cerr.NewInternalError(correlationId, "JSON_CNV_ERR", "Can't convert from JSON to Beacons").WithCause(jsonErr)
		c.SendError(res, req, err)
		return
	}
	result, err := c.controller.UpdateBeacon(
		correlationId,
		&beacon,
	)
	c.SendResult(res, req, result, err)
}

func (c *BeaconsRestService) deleteById(res http.ResponseWriter, req *http.Request) {
	params := req.URL.Query()
	vars := mux.Vars(req)

	beaconId := params.Get("beacon_id")
	if beaconId == "" {
		beaconId = vars["beacon_id"]
	}

	result, err := c.controller.DeleteBeaconById(
		params.Get("correlation_id"),
		beaconId,
	)
	c.SendDeletedResult(res, req, result, err)
}

func (c *BeaconsRestService) Register() {

	c.RegisterRoute(
		"get", "/beacons",
		&cvalid.NewObjectSchema().WithOptionalProperty("skip", cconv.String).
			WithOptionalProperty("take", cconv.String).
			WithOptionalProperty("total", cconv.String).
			WithOptionalProperty("body", cvalid.NewFilterParamsSchema()).Schema,
		c.getPageByFilter,
	)

	c.RegisterRoute(
		"get", "/beacons/{beacon_id}",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("beacon_id", cconv.String).Schema,
		c.getOneById,
	)

	c.RegisterRoute(
		"get", "/beacons/udi/{udi}",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("udi", cconv.String).Schema,
		c.getOneByUdi,
	)

	// Todo: this method shall receive many UDIs! Pass siteid and udi them as query parameters
	c.RegisterRoute(
		"get", "/beacons/calculate_position",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("site_id", cconv.String).
			WithRequiredProperty("beacon_udi", cconv.String).Schema,
		c.calculatePosition,
	)

	c.RegisterRoute(
		"post", "/beacons",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("body", data1.NewBeaconV1Schema()).Schema,
		c.create,
	)

	c.RegisterRoute(
		"put", "/beacons",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("body", data1.NewBeaconV1Schema()).Schema,
		c.update,
	)

	c.RegisterRoute(
		"delete", "/beacons/{beacon_id}",
		&cvalid.NewObjectSchema().
			WithRequiredProperty("beacon_id", cconv.String).Schema,
		c.deleteById,
	)
}

File: service-beacons-dotnet/src/Service/Services/Version1/BeaconsRestServiceV1.cs
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Routing;
using Nov.MaxSamples.Beacons.Data.Version1;
using Nov.MaxSamples.Beacons.Logic;
using PipServices3.Commons.Convert;
using PipServices3.Commons.Data;
using PipServices3.Commons.Refer;
using PipServices3.Commons.Run;
using PipServices3.Rpc.Auth;
using PipServices3.Rpc.Services;
using System;
using System.IO;
using System.Security.Claims;
using System.Threading.Tasks;

namespace Nov.MaxSamples.Beacons.Services.Version1
{
	public class BeaconsRestServiceV1 : RestService
	{
		private IBeaconsController _controller;

		public BeaconsRestServiceV1()
		{
			_baseRoute = "v1/beacons";
			_dependencyResolver.Put("controller", new Descriptor("beacons", "controller", "default", "*", "*"));
		}

		public override void SetReferences(IReferences references)
		{
			base.SetReferences(references);
			_controller = _dependencyResolver.GetOneRequired<IBeaconsController>("controller");
		}

		private async Task GetBeaconsAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.get_beacons";

			try
			{
				var filter = GetFilterParams(request);
				var paging = GetPagingParams(request);

				using var timing = Instrument(correlationId, methodName);
				var result = await _controller.GetBeaconsAsync(
					correlationId,
					filter,
					paging
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task GetBeaconByIdAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.get_beacon_by_id";
			
			try
			{
				var parameters = GetParameters(request);
				var id = parameters.GetAsString("id");

				var result = await _controller.GetBeaconByIdAsync(
					correlationId,
					id
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task GetBeaconByUdiAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.get_beacon_by_udi";

			try
			{
				var parameters = GetParameters(request);
				var udi = parameters.GetAsString("udi");

				var result = await _controller.GetBeaconByUdiAsync(
					correlationId,
					udi
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task CalculatePositionAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.calculate_position";

			try
			{
				var parameters = GetParameters(request);
				var siteId = parameters.GetAsString("site_id");
				string[] udis = ConvertToStringList(parameters.Get("udis"));

				var result = await _controller.CalculatePositionAsync(
					correlationId,
					siteId,
					udis
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task CreateBeaconAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.create_beacon";

			try
			{
				var parameters = GetParameters(request);
				var beacon = ConvertToBeacon(parameters.GetAsObject("beacon"));

				var result = await _controller.CreateBeaconAsync(
					correlationId,
					beacon
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task UpdateBeaconAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.update_beacon";

			try
			{
				var parameters = GetParameters(request);
				var beacon = ConvertToBeacon(parameters.GetAsObject("beacon"));

				var result = await _controller.UpdateBeaconAsync(
					correlationId,
					beacon
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		private async Task DeleteBeaconByIdAsync(HttpRequest request, HttpResponse response, RouteData routeData)
		{
			var correlationId = GetCorrelationId(request);
			var methodName = $"{_baseRoute}.delete_beacon_by_id";

			try
			{
				var parameters = GetParameters(request);
				var id = parameters.GetAsString("id");

				var result = await _controller.DeleteBeaconByIdAsync(
					correlationId,
					id
				);

				await SendResultAsync(response, result);
			}
			catch (Exception ex)
			{
				InstrumentError(correlationId, methodName, ex);
				await SendErrorAsync(response, ex);
			}
		}

		public override void Register()
		{
			RegisterRoute("get", "/", GetBeaconsAsync);
			RegisterRoute("get", "/{id}", GetBeaconByIdAsync);
			RegisterRoute("get", "/udi/{udi}", GetBeaconByUdiAsync);
			RegisterRoute("post", "/", CreateBeaconAsync);
			RegisterRoute("put", "/", UpdateBeaconAsync);
			RegisterRoute("delete", "/{id}", DeleteBeaconByIdAsync);
			RegisterRoute("post", "/calculate_position", CalculatePositionAsync);

			RegisterOpenApiSpecFromResource("beacons_rest_v1.yaml");
		}

		private BeaconV1 ConvertToBeacon(object value)
		{
			return JsonConverter.FromJson<BeaconV1>(JsonConverter.ToJson(value));
		}

		private string[] ConvertToStringList(object value)
		{
			return JsonConverter.FromJson<string[]>(JsonConverter.ToJson(value));
		}
	}
}
